spec/syntax.md

```diff
@@ -91,150 +91,198 @@ The syntax specification takes into account the following design restrictions:
+A **_placeholder_** is an _expression_ that appears inside of a _pattern_
+and which will be replaced during the formatting of the _message_.
 
-A simple _expression_ is a bare variable name:
+An _expression_ begins with U+007B LEFT CURLY BRACKET `{` 
+and ends with U+007D RIGHT CURLY BRACKET `}`.
 
-    {Hello, {$userName}!}
+An _expression_ can appear as a local variable value, as a _selector_, and within a _pattern_.
 
-### Formatting Functions
+> A simple _expression_ containing a variable:
+>
+>>```
+>>    {Hello, {$userName}!}
+>>```
 
-A _function_ is named functionality, possibly with _options_, that format,
-process, or operate on an _operand_ which may be either a _literal_ or a _variable_.
+### Function
 
-For example, a _message_ with a `$date` _variable_ formatted with the `:datetime` _function_:
+A **_function_** is a named modifier in an _expression_.
+A _function_ MAY be followed by zero or more _options_
 
-    {Today is {$date :datetime weekday=long}.}
+>For example, a _message_ with a `$date` _variable_ formatted with the `:datetime` _function_:
+>
+>>```
+>>    {Today is {$date :datetime weekday=long}.}
+>>```
 
-A _message_ with a `$userName` _variable_ formatted with
-the custom `:person` _function_ capable of
-declension (using either a fixed dictionary, algorithmic declension, ML, etc.):
+>A _message_ with a `$userName` _variable_ formatted with
+>the custom `:person` _function_ capable of
+>declension (using either a fixed dictionary, algorithmic declension, ML, etc.):
+>
+>>```
+>>    {Hello, {$userName :person case=vocative}!}
+>>```
 
-    {Hello, {$userName :person case=vocative}!}
+>A _message_ with a `$userObj` _variable_ formatted with
+>the custom `:person` _function_ capable of
+>plucking the first name from the object representing a person:
+>
+>>```
+>>    {Hello, {$userObj :person firstName=long}!}
+>>```
 
-A _message_ with a `$userObj` _variable_ formatted with
-the custom `:person` _function_ capable of
-plucking the first name from the object representing a person:
+_Functions_ can be _standalone_, or can be an _opening element_ or _closing element_.
 
-    {Hello, {$userObj :person firstName=long}!}
+A **_standalone_** _function_ is not expected to be paired with another _function_.
+An **_opening element_** is a _function_ that SHOULD be paired with a _closing function_.
+A **_closing element_** is a _function_ that SHOULD be paired with an _opening function_.
 
-A message with two markup-like _functions_, `button` and `link`,
-which the runtime can use to construct a document tree structure for a UI framework:
+An _opening element_ MAY be present in a message without a corresponding _closing element_,
+and vice versa.
 
-    {{+button}Submit{-button} or {+link}cancel{-link}.}
+>A message with two markup-like _functions_, `button` and `link`,
+>which the runtime can use to construct a document tree structure for a UI framework:
+>
+>>```
+>>    {{+button}Submit{-button} or {+link}cancel{-link}.}
+>>```
```

Comment:
With the triple backticking, the four-space indentations should also be removed. This also applies to other places where this change is applied.